contracts/cw-storage/src/contract.rs

```diff
@@ -116,6 +120,78 @@ pub mod execute {
 
         Ok(res)
     }
+
+    pub fn pin_object(
+        deps: DepsMut,
+        info: MessageInfo,
+        object_id: ObjectId,
+    ) -> Result<Response, ContractError> {
+        if pins().has(deps.storage, (object_id.clone(), info.sender.clone())) {
+            return Ok(Response::new()
+                .add_attribute("action", "pin_object")
+                .add_attribute("id", object_id));
+        }
+
+        let bucket = BUCKET.load(deps.storage)?;
+
+        let o = objects().update(
+            deps.storage,
+            object_id.clone(),
+            |o| -> Result<Object, StdError> {
+                o.map(|mut e: Object| -> Object {
+                    e.pin_count += Uint128::one();
+                    e
+                })
+                .ok_or_else(|| StdError::not_found(type_name::<Object>()))
+            },
+        )?;
+
+        match bucket.limits {
+            Limits {
+                max_object_pins: Some(max),
+                ..
+            } if max < o.pin_count => {
+                Err(BucketError::MaxObjectPinsLimitExceeded(o.pin_count, max).into())
+            }
+            _ => {
+                pins().save(
+                    deps.storage,
+                    (object_id.clone(), info.sender.clone()),
+                    &Pin {
+                        id: object_id.clone(),
+                        address: info.sender,
+                    },
+                )?;
+                Ok(Response::new()
+                    .add_attribute("action", "pin_object")
+                    .add_attribute("id", object_id))
```

Comment:
The response code is duplicated, we can declare it first and use it as it is done in the unpin 😉